cas-server-support-osf/src/main/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction.java

```diff
@@ -269,65 +281,102 @@ protected Credential constructCredentialsFromRequest(final RequestContext contex
         // The call is made to trust these headers only because we assume the Apache Shibboleth Service Provider module
         // rejects any conflicting / forged headers.
         final String shibbolethSession = request.getHeader(SHIBBOLETH_SESSION_HEADER);
+
         if (StringUtils.hasText(shibbolethSession)) {
+            // SAML-Shibboleth based institution login
             final String remoteUser = request.getHeader(REMOTE_USER);
             if (StringUtils.isEmpty(remoteUser)) {
                 logger.error("Invalid Remote User specified as Empty");
                 throw new RemoteUserFailedLoginException("Invalid Remote User specified as Empty");
+            } else {
+                logger.info("Remote User from HttpServletRequest '{}'", remoteUser);
+                credential.setRemotePrincipal(Boolean.TRUE);
             }
-
-            logger.info("Remote User from HttpServletRequest '{}'", remoteUser);
-            credential.setRemotePrincipal(Boolean.TRUE);
-
             for (final String headerName : Collections.list(request.getHeaderNames())) {
                 if (headerName.startsWith(ATTRIBUTE_PREFIX)) {
                     final String headerValue = request.getHeader(headerName);
                     logger.debug("Remote User [{}] Auth Header '{}': '{}'", remoteUser, headerName, headerValue);
-
-                    credential.getAuthenticationHeaders().put(
+                    credential.getDelegationAttributes().put(
                             headerName.substring(ATTRIBUTE_PREFIX.length()),
                             headerValue
                     );
                 }
             }
-
+            credential.getDelegationAttributes().put("Delegation-Protocol", PROTOCOL_SAML);
+            credential.setDelegationProtocol(PROTOCOL_SAML);
             // Notify the OSF of the remote principal authentication.
-            final PrincipalAuthenticationResult remoteUserInfo = notifyRemotePrincipalAuthenticated(credential);
+            final PrincipalAuthenticationResult remoteUserInfo
+                    = notifyRemotePrincipalAuthenticated(credential, null);
             credential.setUsername(remoteUserInfo.getUsername());
             credential.setInstitutionId(remoteUserInfo.getInstitutionId());
-
+            return credential;
+        } else if (PROTOCOL_CAS.equals(credential.getDelegationProtocol())) {
```

Comment:
From flow action `clientAction` to `remoteAuthenticate` during `CAS_PAC4J` institution login, the `credential` contains the protocol information and institution id.